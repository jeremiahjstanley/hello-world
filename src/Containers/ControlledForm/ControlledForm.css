form {
  display: inline-block;
  text-align: left;
  max-width: 40rem;
}

form p,input, select, button  {
  font-family: Avenir Next;
  font-size: 1.25rem;
  font-weight: 300;
}

form, p {
  display: inline-block;
  padding-right: 1rem;
}

input, select  {
  background-color: #FFFFFF;
  border: none;
  border-bottom: 1px solid #000000;
  max-width: 25rem;
}

input::placeholder {
  color: #000000;
}

button {
  background-color: #000000;
  color: #FFFFFF;
  margin-top: 1.5rem;
  width: 100%;
}

button:disabled {
  opacity: .05
}

.loading-text {
  margin: 0;
}