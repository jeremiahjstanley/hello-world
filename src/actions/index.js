export const setLocation = (location) => ({
  type: 'SET_LOCATION',
  location
});

export const setDataSet = (dataSet) => ({
  type: 'SET_DATASET',
  dataSet
});

export const clearDataSet = () => ({
  type: 'CLEAR_DATASET',
});

export const setDataBase = (dataBase) => ({
  type: 'SET_DATABASE',
  dataBase
});

export const fetchLocationDataSuccess = (locationData) => ({
  type: 'FETCH_LOCATION_DATA_SUCCESS',
  locationData
});

export const hasErrored = (status) => ({
  type: 'HAS_ERRORED',
  hasErrored: status
});

export const madeSearch = (status) => ({
  type: 'MADE_SEARCH',
  madeSearch: status
});


