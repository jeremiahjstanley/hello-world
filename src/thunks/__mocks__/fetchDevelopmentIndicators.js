export const fetchDevelopmentIndicators = jest.fn().mockImplementation(() => {
  return {};
});