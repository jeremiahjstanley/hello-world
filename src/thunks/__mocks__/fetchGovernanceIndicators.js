export const fetchGovernanceIndicators = jest.fn().mockImplementation(() => {
  return {};
});