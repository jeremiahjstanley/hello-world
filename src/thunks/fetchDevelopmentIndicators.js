import { fetchLocationDataSuccess } from '../actions';
import { dataCleaner } from '../helper/dataCleaner';
import { hasErrored } from '../actions';
import { key } from '../helper/apiKey';

export const fetchDevelopmentIndicators = (locations, dataSet, dataBase) => {

    return async (dispatch) => {
        const unresolvedDevelopmentIndicators = locations.map(async location => {
            const isoAlpha3 = location.alpha_3
            try {
              const url = `https://www.quandl.com/api/v3/datasets/${dataBase}/${isoAlpha3}_${dataSet}.json?api_key=${key}`;
              const response = await fetch(url);
              if (!response.ok) {
                throw Error(response.statusText);
              }
              return await response.json();
            } catch (error) {
              dispatch(hasErrored(true));
            }
        });

        const developmentIndicators = await Promise.all(unresolvedDevelopmentIndicators);
        const cleanDevelopmentIndicators = developmentIndicators.map(indicator => {
            return dataCleaner(indicator);
        });
        
        await dispatch(fetchLocationDataSuccess(cleanDevelopmentIndicators))   
    };
};