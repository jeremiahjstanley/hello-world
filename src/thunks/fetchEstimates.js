import { hasErrored } from '../actions';
import { key } from '../helper/apiKey';

export const fetchEstimates = (isoAlpha3, dataSet, dataBase) => {
  return async (dispatch) => {
    try {
      const url = `https://www.quandl.com/api/v3/datasets/${dataBase}/${isoAlpha3}_${dataSet}_EST.json?api_key=${key}`;
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return await response.json();
    } catch (error) {
      dispatch(hasErrored(true));
    }
  }
};