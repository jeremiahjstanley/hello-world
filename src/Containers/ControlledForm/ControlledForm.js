import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setInitialLocation } from '../../thunks/setInitialLocation';
import { dataMetrics } from '../../helper/dataMetrics';

export class ControlledForm extends Component {
  constructor() {
    super()

    this.state = {
      initialLocation: '',
      initialMetric: dataMetrics[0]
    }
  }

  handleChange = (event) => {
    const initialLocation = event.target.value;
    this.setState({ initialLocation });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const location = this.state.initialLocation;
    const dataSet = this.state.initialMetric.dataset_code;
    const dataBase = this.state.initialMetric.database_code;
    this.props.selectInitialLocation(location, dataSet, dataBase);
  }

  selectMetric = (event) => {
    const initialMetric = dataMetrics.find(metric => event.target.value === metric.name);
    this.setState({ initialMetric })
  }

  render() {

    const options = dataMetrics.map((metric, index) => (
      <option key={index}>
        {metric.name}
      </option>
    ))

    return (
      <form onSubmit={this.handleSubmit}>
       <input
         value={this.state.initialLocation}
         type='text'
         placeholder='enter a country'
         onChange={this.handleChange}
       /> 
       <select onChange={this.selectMetric}>
        { options }
       </select>
       <button>
         Search
       </button>
     </form>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  selectInitialLocation: (initialLocation, dataSet, dataBase) => dispatch(setInitialLocation(initialLocation, dataSet, dataBase))
})

export default connect(null, mapDispatchToProps)(ControlledForm);