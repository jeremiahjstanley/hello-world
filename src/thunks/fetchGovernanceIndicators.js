import { fetchLocationDataSuccess } from '../actions';
import { dataCleaner } from '../helper/dataCleaner';
import { fetchEstimates } from './fetchEstimates';
import { fetchNumberOfSources } from './fetchNumberOfSources'; 
import { fetchPercentileRank } from './fetchPercentileRank';
import { fetchStandardError } from './fetchStandardError';

export const fetchGovernanceIndicators = (locations, dataSet, dataBase) => {
    return async (dispatch) => {
        const unresolvedGovernanceIndicators = locations.map(async location => {
            const alpha3 = location.alpha_3
            const estimates = await dispatch(fetchEstimates(alpha3, dataSet, dataBase));
            const numberOfSources = await dispatch(fetchNumberOfSources(alpha3, dataSet, dataBase)); 
            const percentileRank = await dispatch(fetchPercentileRank(alpha3, dataSet, dataBase));
            const standardError = await dispatch(fetchStandardError(alpha3, dataSet, dataBase));
            const cleanEstimates = dataCleaner(estimates);
            const cleanNumberOfSources = dataCleaner(numberOfSources);
            const cleanPercentileRank = dataCleaner(percentileRank);
            const cleanStandardError = dataCleaner(standardError);
            return {cleanEstimates, cleanNumberOfSources, cleanPercentileRank, cleanStandardError};
        });
        const governanceIndicators = await Promise.all(unresolvedGovernanceIndicators)
        await dispatch(fetchLocationDataSuccess(governanceIndicators))     
    };
};