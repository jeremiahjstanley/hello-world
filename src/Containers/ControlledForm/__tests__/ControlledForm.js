import React from 'react';
import { shallow } from 'enzyme';
import { ControlledForm, mapStateToProps, mapDispatchToProps } from '../ControlledForm';
import { setLocation, setDataBase, setDataSet, madeSearch } from '../../../actions';
import { fetchGovernanceIndicators }from '../../../thunks/fetchGovernanceIndicators';
import { fetchDevelopmentIndicators } from '../../../thunks/fetchDevelopmentIndicators';

jest.mock('../../../thunks/fetchGovernanceIndicators');
jest.mock('../../../thunks/fetchDevelopmentIndicators');


describe('ControlledForm', () => {

	let mockInput;
	let mockEvent;
	let mockLocation;
	let mockHistory;
  let mockDataBase;
  let mockDataSet;
  let mockFetchDevelopmentIndicators;
  let mockFetchGovernanceIndicators;
  let mockSelectDataSet;
  let mockSelectDataBase;
  let mockSelectLocation;
  let mockMakeSearch;
  let additionalLocation;
  let wrapper;

  beforeEach(() => {

  	mockInput = { target: { value: 'Burkina Faso'} };

   	mockEvent = { preventDefault: jest.fn() };

  	mockHistory = [];

  	mockLocation = [{
      name:'Burkina Faso',
      alpha_2:'BF',
      alpha_3:'BFA',
      country_code:'854',
      iso_3166_2:'ISO 3166-2: BF',
      region:'Africa',
      sub_region:'Sub-Saharan Africa',
      intermediate_region:'Western Africa',
      region_code:'002',sub_region_code:'202',
      intermediate_region_code:'011'
    }];

    mockDataBase = 'WWDI';

    mockDataSet = 'SM_POP_REFG';

	  mockFetchDevelopmentIndicators = jest.fn();

	  mockFetchGovernanceIndicators = jest.fn();

	  mockSelectDataBase = jest.fn();

	  mockSelectDataSet = jest.fn();

	  mockSelectLocation = jest.fn();

    mockMakeSearch = jest.fn()

    wrapper = shallow(
    	<ControlledForm 
    		dataBase={mockDataBase}
    		dataSet={mockDataSet}
    		fetchGovernanceIndicators={mockFetchGovernanceIndicators}
    		fetchDevelopmentIndicators={mockFetchDevelopmentIndicators}
    		selectDataBase={mockSelectDataBase}
    		selectDataSet={mockSelectDataSet}
    		selectLocation={mockSelectLocation}
        makeSearch={mockMakeSearch}
    		history={mockHistory}
    	/>
  	);

  });

	it('should match the snapshot', () => {

		expect(wrapper).toMatchSnapshot();

	});

  it('should have a default state with a dataBase and dataSet object', () => {

    const expected = {
      dataBase: {
        name: 'Worldwide Governance Indicators', 
        database_code: 'WWGI'
      },
      dataSet: {
        name: 'Voice and Accountability', 
        dataset_code: 'VA'
      }
    };

    const results = wrapper.state();

    expect(results).toEqual(expected);

  });

  it('should call selectDataBase on page load', () => {

    const expected = wrapper.state().dataBase;

    expect(mockSelectDataBase).toHaveBeenCalledWith(expected);

  });

  it('should call selectDataSet on page load', () => {

    const expected = wrapper.state().dataSet;

    expect(mockSelectDataSet).toHaveBeenCalledWith(expected);

  });

  it('should update the state when handleChange is invoked', () => {

  	const expected = {
      name:'Burkina Faso',
      alpha_2:'BF',
      alpha_3:'BFA',
      country_code:'854',
      iso_3166_2:'ISO 3166-2: BF',
      region:'Africa',
      sub_region:'Sub-Saharan Africa',
      intermediate_region:'Western Africa',
      region_code:'002',sub_region_code:'202',
      intermediate_region_code:'011'
    }

    wrapper.instance().handleChange(mockInput);

    const results = wrapper.state('location');

    expect(results).toEqual(expected);

  });

   it('should invoke fetchGovernanceIndicators when the form is submitted with a database_code of WWDI', () => {

	  mockDataBase = 'WWGI'; 

    wrapper = shallow(
    	<ControlledForm 
    		dataBase={mockDataBase}
    		dataSet={mockDataSet}
    		fetchGovernanceIndicators={mockFetchGovernanceIndicators}
    		fetchDevelopmentIndicators={mockFetchDevelopmentIndicators}
    		selectDataBase={mockSelectDataBase}
    		selectDataSet={mockSelectDataSet}
    		selectLocation={mockSelectLocation}
        makeSearch={mockMakeSearch}
    		history={mockHistory}
    	/>
    );

   	wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockFetchGovernanceIndicators).toHaveBeenCalledWith(mockLocation, mockDataSet, mockDataBase);

  });

   it('should invoke fetchDevelopmentIndicators when the form is submitted with a database_code other than WWGI', () => {

   	wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockFetchDevelopmentIndicators).toHaveBeenCalledWith(mockLocation, mockDataSet, mockDataBase);

  });

   it('should direct the user to the stats page of the application', () => {

   	wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockHistory).toEqual(['/stats']);

  });

	describe('mapStateToProps', () => {

		let mockState;
		let mockDataBase;
		let mockDataSet;

		beforeEach(() => {

			mockDataBase = {
    		name: 'Worldwide Governance Indicators',
    		database_code: 'WWGI'
    	}

    	mockDataSet = {
        name: 'Voice and Accountability', 
        dataset_code: 'VA'
      }

			mockState = {
				dataBase: mockDataBase,
				dataSet: mockDataSet,
			};

		});

		it('should return an object with dataBase and dataSet key', () => {

			const expected = {
				dataBase: mockDataBase.database_code,
				dataSet: mockDataSet.dataset_code,
			};

			const mappedProps = mapStateToProps(mockState);

			expect(mappedProps).toEqual(expected);

		});

	});

	describe('mapDispatchToProps', () => {

		let actionToDispatch;
		let mockDispatch;
		let mockDataBase;
		let mockDataSet

		beforeEach(() => {

			mockDispatch = jest.fn();

			mockDataBase = {
    		name: 'Worldwide Governance Indicators',
    		database_code: 'WWGI'
    	};

    	mockDataSet = {
        name: 'Voice and Accountability', 
        dataset_code: 'VA'
      };

		});

		it('calls dispatch with the fetchDevelopmentIndicators action when called', () => {

	  	mockDataBase = 'WWGI'; 

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = fetchDevelopmentIndicators(mockLocation, mockDataBase, mockDataSet);

			mappedProps.fetchDevelopmentIndicators(mockLocation, mockDataBase, mockDataSet);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});

		it('calls dispatch with the fetchDevelopmentIndicators action when called', () => {

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = fetchGovernanceIndicators(mockLocation, mockDataBase, mockDataSet);

			mappedProps.fetchGovernanceIndicators(mockLocation, mockDataBase, mockDataSet);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});


		it('calls dispatch with the setDataBase action when called', () => {

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = setDataBase(mockDataBase);

			mappedProps.selectDataBase(mockDataBase);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});

		it('calls dispatch with the setDataSet action when called', () => {

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = setDataSet(mockDataSet);

			mappedProps.selectDataSet(mockDataSet);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});

		it('calls dispatch with the setLocation action when called', () => {

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = setLocation(mockLocation);

			mappedProps.selectLocation(mockLocation);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});

    it('calls dispatch with the madeSearch action when called', () => {

      const mappedProps = mapDispatchToProps(mockDispatch);

      actionToDispatch = madeSearch(mockLocation);

      mappedProps.makeSearch(mockLocation);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

    });
		
	});

});