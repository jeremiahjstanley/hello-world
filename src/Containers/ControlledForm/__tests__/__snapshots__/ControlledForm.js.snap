// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledForm should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ControlledForm
    dataBase="WWDI"
    dataSet="SM_POP_REFG"
    fetchDevelopmentIndicators={[Function]}
    fetchGovernanceIndicators={[Function]}
    history={Array []}
    selectDataBase={[Function]}
    selectDataSet={[Function]}
    selectLocation={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <Link
                    className="explanation-button"
                    replace={false}
                    to="/why"
          >
                    ▼ 
                    <span>
                              why?
                    </span>
          </Link>
          <p>
                    I'm curious about...
          </p>
          <input
                    list="countries"
                    onChange={[Function]}
                    placeholder="Enter a Country"
                    type="text"
          />
          <datalist
                    id="countries"
          >
                    <option>
                              Afghanistan
                    </option>
                    <option>
                               Åland Islands
                    </option>
                    <option>
                              Albania
                    </option>
                    <option>
                              Algeria
                    </option>
          </datalist>
</div>,
        <div>
          <p>
                    and their...
          </p>
          <Connect(DataBaseSelectField) />
</div>,
        <div>
          <p>
                    but specifically, 
          </p>
          <Connect(DataSetSelectField) />
</div>,
        <button
          disabled={true}
>
          Search
</button>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              className="explanation-button"
              replace={false}
              to="/why"
>
              ▼ 
              <span>
                            why?
              </span>
</Link>,
            <p>
              I'm curious about...
</p>,
            <input
              list="countries"
              onChange={[Function]}
              placeholder="Enter a Country"
              type="text"
/>,
            <datalist
              id="countries"
>
              <option>
                            Afghanistan
              </option>
              <option>
                             Åland Islands
              </option>
              <option>
                            Albania
              </option>
              <option>
                            Algeria
              </option>
</datalist>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "▼ ",
                <span>
                  why?
</span>,
              ],
              "className": "explanation-button",
              "replace": false,
              "to": "/why",
            },
            "ref": null,
            "rendered": Array [
              "▼ ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "why?",
                },
                "ref": null,
                "rendered": "why?",
                "type": "span",
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "I'm curious about...",
            },
            "ref": null,
            "rendered": "I'm curious about...",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "list": "countries",
              "onChange": [Function],
              "placeholder": "Enter a Country",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option>
                  Afghanistan
</option>,
                <option>
                   Åland Islands
</option>,
                <option>
                  Albania
</option>,
                <option>
                  Algeria
</option>,
              ],
              "id": "countries",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "host",
                "props": Object {
                  "children": "Afghanistan",
                },
                "ref": null,
                "rendered": "Afghanistan",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": " Åland Islands",
                },
                "ref": null,
                "rendered": " Åland Islands",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "2",
                "nodeType": "host",
                "props": Object {
                  "children": "Albania",
                },
                "ref": null,
                "rendered": "Albania",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "3",
                "nodeType": "host",
                "props": Object {
                  "children": "Algeria",
                },
                "ref": null,
                "rendered": "Algeria",
                "type": "option",
              },
            ],
            "type": "datalist",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              and their...
</p>,
            <Connect(DataBaseSelectField) />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "and their...",
            },
            "ref": null,
            "rendered": "and their...",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              but specifically, 
</p>,
            <Connect(DataSetSelectField) />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "but specifically, ",
            },
            "ref": null,
            "rendered": "but specifically, ",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Search",
          "disabled": true,
        },
        "ref": null,
        "rendered": "Search",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <Link
                        className="explanation-button"
                        replace={false}
                        to="/why"
            >
                        ▼ 
                        <span>
                                    why?
                        </span>
            </Link>
            <p>
                        I'm curious about...
            </p>
            <input
                        list="countries"
                        onChange={[Function]}
                        placeholder="Enter a Country"
                        type="text"
            />
            <datalist
                        id="countries"
            >
                        <option>
                                    Afghanistan
                        </option>
                        <option>
                                     Åland Islands
                        </option>
                        <option>
                                    Albania
                        </option>
                        <option>
                                    Algeria
                        </option>
            </datalist>
</div>,
          <div>
            <p>
                        and their...
            </p>
            <Connect(DataBaseSelectField) />
</div>,
          <div>
            <p>
                        but specifically, 
            </p>
            <Connect(DataSetSelectField) />
</div>,
          <button
            disabled={true}
>
            Search
</button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="explanation-button"
                replace={false}
                to="/why"
>
                ▼ 
                <span>
                                why?
                </span>
</Link>,
              <p>
                I'm curious about...
</p>,
              <input
                list="countries"
                onChange={[Function]}
                placeholder="Enter a Country"
                type="text"
/>,
              <datalist
                id="countries"
>
                <option>
                                Afghanistan
                </option>
                <option>
                                 Åland Islands
                </option>
                <option>
                                Albania
                </option>
                <option>
                                Algeria
                </option>
</datalist>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "▼ ",
                  <span>
                    why?
</span>,
                ],
                "className": "explanation-button",
                "replace": false,
                "to": "/why",
              },
              "ref": null,
              "rendered": Array [
                "▼ ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "why?",
                  },
                  "ref": null,
                  "rendered": "why?",
                  "type": "span",
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "I'm curious about...",
              },
              "ref": null,
              "rendered": "I'm curious about...",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "list": "countries",
                "onChange": [Function],
                "placeholder": "Enter a Country",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option>
                    Afghanistan
</option>,
                  <option>
                     Åland Islands
</option>,
                  <option>
                    Albania
</option>,
                  <option>
                    Algeria
</option>,
                ],
                "id": "countries",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "0",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Afghanistan",
                  },
                  "ref": null,
                  "rendered": "Afghanistan",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "host",
                  "props": Object {
                    "children": " Åland Islands",
                  },
                  "ref": null,
                  "rendered": " Åland Islands",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "2",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Albania",
                  },
                  "ref": null,
                  "rendered": "Albania",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "3",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Algeria",
                  },
                  "ref": null,
                  "rendered": "Algeria",
                  "type": "option",
                },
              ],
              "type": "datalist",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                and their...
</p>,
              <Connect(DataBaseSelectField) />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "and their...",
              },
              "ref": null,
              "rendered": "and their...",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                but specifically, 
</p>,
              <Connect(DataSetSelectField) />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "but specifically, ",
              },
              "ref": null,
              "rendered": "but specifically, ",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search",
            "disabled": true,
          },
          "ref": null,
          "rendered": "Search",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
